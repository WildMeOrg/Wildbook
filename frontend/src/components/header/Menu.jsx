import React, { useState } from "react";
import { Nav, NavDropdown } from "react-bootstrap";
import { FormattedMessage } from "react-intl";
import DownIcon from "../svg/DownIcon";
import RightIcon from "../svg/RightIcon";
import { authenticatedMenu } from "../../constants/navMenu";

export default function Menu({ username }) {
  const [dropdownColor, setDropdownColor] = useState("transparent");

  const [dropdownShows, setDropdownShows] = useState({});
  const [dropdownBorder, setDropdownBorder] = useState("2px solid transparent");

  const handleMouseEnterLeave = (id, isEnter) => {
    setDropdownShows((prev) => ({ ...prev, [id]: isEnter ? true : false }));
    setDropdownBorder((prev) => ({
      ...prev,
      [id]: isEnter ? "2px solid white" : "2px solid transparent",
    }));
  };

  return (
    <>
      {authenticatedMenu(username).map((item, idx) => (
        <Nav className="me-auto" key={idx}>
          <NavDropdown
            className="header-dropdown"
            title={
              <span style={{ color: "white" }}>
                <FormattedMessage id={Object.keys(item)[0].toUpperCase()} />{" "}
                <DownIcon color={"white"} />
              </span>
            }
            id={`basic-nav-dropdown${idx}`}
            style={{
              color: "white",
              boxSizing: "border-box",
              paddingLeft: 5,
              paddingRight: 5,
              borderBottom:
                dropdownBorder[`dropdown${idx + 1}`] || "2px solid transparent",
            }}
            onMouseEnter={() =>
              handleMouseEnterLeave(`dropdown${idx + 1}`, true)
            }
            onMouseLeave={() =>
              handleMouseEnterLeave(`dropdown${idx + 1}`, false)
            }
            show={dropdownShows[`dropdown${idx + 1}`]}
          >
            {Object.values(item)[0].map((subItem) => {
              return subItem.sub ? (
                <NavDropdown
                  className="header-dropdown"
                  title={
                    <a
                      style={{
                        color: "black",
                        fontSize: "0.9rem",
                        textDecoration: "none",
                      }}
                      onClick={(e) => {
                        e.stopPropagation();
                        e.preventDefault();
                        window.location.href = subItem.href;
                      }}
                      href={subItem.href}
                    >
                      {subItem.name}
                      <span style={{ paddingLeft: "34px" }}>
                        <RightIcon />
                      </span>
                    </a>
                  }
                  drop="end"
                  style={{
                    paddingLeft: 8,
                    fontSize: "0.9rem",
                    backgroundColor: dropdownColor,
                  }}
                  onMouseEnter={() => {
                    setDropdownColor("#CCF0FF");
                    handleMouseEnterLeave(`dropdown7`, true);
                  }}
                  onMouseLeave={() => {
                    setDropdownColor("white");
                    handleMouseEnterLeave(`dropdown7`, false);
                  }}
                  show={dropdownShows[`dropdown7`]}
                >
                  {subItem.sub.map((sub) => {
                    return (
                      <NavDropdown.Item
                        key={sub?.name}
                        href={sub.href}
                        style={{ color: "black", fontSize: "0.9rem" }}
                      >
                        {sub.name}
                      </NavDropdown.Item>
                    );
                  })}
                </NavDropdown>
              ) : (
                <NavDropdown.Item
                  href={subItem.href}
                  style={{ color: "black", fontSize: "0.9rem" }}
                >
                  {subItem.name}
                </NavDropdown.Item>
              );
            })}
          </NavDropdown>
        </Nav>
      ))}
    </>
  );
}
