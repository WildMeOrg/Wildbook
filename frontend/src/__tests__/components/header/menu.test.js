import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import Menu from "../../../components/header/Menu";
import { IntlProvider } from "react-intl";
import "@testing-library/jest-dom";

jest.mock("../../../constants/navMenu", () => ({
  authenticatedMenu: () => [
    {
      Submit: [
        {
          name: "Report an Encounter",
          href: "/report",
          sub: [
            { name: "SubItem 1", href: "/sub-item-1" },
            { name: "SubItem 2", href: "/sub-item-2" },
          ],
        },
      ],
    },
  ],
}));

test('renders "Submit" in the menu', () => {
  render(
    <IntlProvider locale="en">
      <Menu username="tomcat" />
    </IntlProvider>,
  );
  expect(screen.getByText("SUBMIT")).toBeInTheDocument();
});

test("opens dropdown on hover", async () => {
  render(
    <IntlProvider locale="en">
      <Menu username="testuser" />
    </IntlProvider>,
  );
  const menuItem = screen.getAllByRole("button")[0];
  fireEvent.mouseEnter(menuItem);
  await waitFor(() => {
    expect(menuItem.getAttribute("aria-expanded")).toBe("true");
  });
});

test("closes dropdown on mouse leave", async () => {
  render(
    <IntlProvider locale="en">
      <Menu username="testuser" />
    </IntlProvider>,
  );
  const menuItem = screen.getAllByRole("button")[0];
  fireEvent.mouseEnter(menuItem);
  await waitFor(() =>
    expect(menuItem.getAttribute("aria-expanded")).toBe("true"),
  );
  fireEvent.mouseLeave(menuItem);
  await waitFor(() =>
    expect(menuItem.getAttribute("aria-expanded")).toBe("false"),
  );
});

test("renders sub-menu items when parent is hovered", async () => {
  render(
    <IntlProvider locale="en">
      <Menu username="testuser" />
    </IntlProvider>,
  );
  const parentMenu1 = screen.getByText("SUBMIT");
  fireEvent.mouseEnter(parentMenu1);
  const parentMenu2 = screen.getByText("Report an Encounter");
  fireEvent.mouseEnter(parentMenu2);

  await waitFor(() => {
    expect(screen.getByText("SubItem 1")).toBeInTheDocument();
    expect(screen.getByText("SubItem 2")).toBeInTheDocument();
    const link = screen.getByText("SubItem 1");
    expect(link).toHaveAttribute("href", "/sub-item-1");
  });
});

test("sets dropdown color to white on mouse leave", async () => {
  render(
    <IntlProvider locale="en">
      <Menu username="testuser" />
    </IntlProvider>,
  );

  const parentMenu1 = screen.getByText("SUBMIT");
  fireEvent.mouseEnter(parentMenu1);
  const parentMenu2 = screen.getByText("Report an Encounter");
  fireEvent.mouseEnter(parentMenu2);

  const subMenu = screen.getByText("SubItem 1");

  fireEvent.mouseEnter(subMenu);
  fireEvent.mouseLeave(subMenu);

  await waitFor(() => {
    expect(subMenu.parentElement).not.toHaveClass("show");
  });
});
