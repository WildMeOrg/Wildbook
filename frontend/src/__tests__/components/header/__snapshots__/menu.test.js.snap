// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders "Submit" in the menu and matches snapshot 1`] = `
<span
  style="color: white;"
>
  SUBMIT
   
  <svg
    class="bi bi-chevron-down"
    fill="white"
    height="12"
    viewBox="0 0 16 16"
    width="12"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708"
      fill-rule="evenodd"
    />
  </svg>
</span>
`;
